src/packed-pack.ts

```diff
@@ -283,19 +280,101 @@ export class PackContent {
                 reader.offset = relOffset;
             }
 
-            // n-byte type and length (3-bit type, (n-1)*7+4-bit length) compressed data
-            // https://git-scm.com/docs/pack-format#_size_encoding
-            const int = readVarIntLE(reader);
-            const btype = ((int >> 4) & 0b0111) as PackedType;
-            const length = ((int >> 7) << 4) | (int & 0b1111);
+            const objectPrelude = await this.readObjectPrelude(offsetBase, reader);
+            const { btype, unpackedSize } = objectPrelude;
 
             objectsByType[btype].count++;
             objectsByType[btype].size += nextOffset - offset;
-            objectsByType[btype].unpackedSize += length;
+            objectsByType[btype].unpackedSize += unpackedSize;
+
+            if (isOffsetDeltaPrelude(objectPrelude)) {
+                objectsByType[btype].unpackedRestoredSize += objectPrelude.unpackedRestoredSize;
+            }
         }
 
         return objectsByType.filter((stat) => stat.count > 0);
     }
+
+    private read(buffer: Buffer, offset: number) {
+        return this.fh.read(buffer, 0, buffer.byteLength, offset);
+    }
+
+    private async readObjectPreludeFromFile(
+        offset: number
+    ): Promise<[BufferCursor, ObjectPrelude]> {
+        const header =
+            reuseBufferCount > 0 ? buffers[--reuseBufferCount] : Buffer.allocUnsafe(4096);
+
+        await this.read(header, offset);
+        const reader = new BufferCursor(header);
+
+        return [reader, await this.readObjectPrelude(offset, reader)];
```

Comment:
Avoid creating a temporary structures (objects, arrays etc.). So this array will be disposed after a destruction on a receiver side. Creating a lot of temporary objects (and we have a lot of objects) leads to a pressure on GC (GC trashing), that's bad for performance too. As a solution, `reader` might be stored in prelude.